[#4091] Migration paths documentation for serializer/converter (#4227)
AxonFramework/AxonFramework: docs/reference-guide/modules/migration/pages/paths/serializers.adoc

The `org.axonframework.serialization.Serializer` interface and all its specific implementations (like `JacksonSerializer`, `XStreamSerializer`, etc.) have been removed. They are replaced by the `org.axonframework.conversion.Converter` interface.

In most cases, this is a straightforward replacement:

* Instead of `JacksonSerializer`, use `JacksonConverter`.
* Instead of `XStreamSerializer`, see the section on its removal below.

The `Converter` interface focuses on converting data between different content types (like `byte[]`, `JsonNode`, or specific Java objects) without the heavy baggage of the previous serialization API.

== Removal of XStreamSerializer

The `XStreamSerializer` has been removed in Axon Framework 5. XStream is nearing its end-of-life, and the framework has moved towards Jackson as the default and recommended conversion library.

For users who require an XML-based format, it is recommended to use the `JacksonConverter` configured with an `XmlMapper` from the `jackson-dataformat-xml` artifact.

[source,java]
----
// Configuring Jackson for XML conversion
XmlMapper xmlMapper = new XmlMapper();
Converter xmlConverter = new JacksonConverter(xmlMapper);
----

> == Message and Event Converter Types

Reviewer: ```suggestion
== Message and event converter types
```
or
```suggestion
== Message and Event converter types
```

fixes the vale sentence-style capitalization error (`Event` is a known term in the vale vocabulary already)

Infrastructure components now explicitly require these more specific types where appropriate. For instance, the `EventStorageEngine` now expects an `EventConverter`.

== Declarative Configuration (Non-Spring)

When using the `MessagingConfigurer` in a non-Spring environment, the configuration has shifted from the `Configurer` to a more modular approach.

[source,java]
----
// AF4
public void configure(Configurer configurer) {
    configurer.configureSerializer(c -> JacksonSerializer.defaultSerializer())
              .configureMessageSerializer(c -> JacksonSerializer.defaultSerializer())
              .configureEventSerializer(c -> JacksonSerializer.defaultSerializer());
}

// AF5
public void configure(MessagingConfigurer configurer) {
    configurer.componentRegistry(registry -> {
        Converter generalConverter = new JacksonConverter();
        registry.registerComponent(Converter.class, c -> generalConverter);
        registry.registerComponent(MessageConverter.class, c -> new DelegatingMessageConverter(generalConverter));
        registry.registerComponent(EventConverter.class, c -> new DelegatingEventConverter(generalConverter));
    });
}
----

> == Spring Configuration

Reviewer: ```suggestion
== Spring configuration
```

fixes the vale sentence-style capitalization error

== Removal of `XStreamSerializer`

The `XStreamSerializer` has been removed in Axon Framework 5. XStream is nearing its end-of-life, and the framework has moved towards Jackson as the default and recommended conversion library.

For users who require an XML-based format, it is recommended to use the `JacksonConverter` configured with an `XmlMapper` from the `jackson-dataformat-xml` artifact.

[source,java]
----
// Configuring Jackson for XML conversion
XmlMapper xmlMapper = new XmlMapper();
Converter xmlConverter = new JacksonConverter(xmlMapper);
----

[WARNING]
.Incompatibility between Jackson and XStream
====
Using the `XmlMapper` as stated above provides a valid solution for `XML` as serialization format. However, it is not safe to assume  that the resulting XML format will be compatible with the one produced by the `XStreamSerializer`.

If you have used the `XStreamSerializer` in for serialization of your events, you should expect to have to reprocess your event store with the new converter, as the format will likely be different.

> xref:paths/index.adoc#missing-a-migration-path[Contact the AxonIQ support team] for assistance with this migration step.

Reviewer 2: ```suggestion
xref:paths/index.adoc#missing-a-migration-path[Contact the Axon Framework team] for assistance with this migration step.
```
